### ISSUE TITLE
Clangd crashes, no way to restart

### ISSUE THREAD

**Issue by Leon0402** [clangd/vscode-clangd#70]
For some reason, clangd crashed a few times and I got the following message: "The Clang Language Server server crashed 5 times in the last 3 minutes. The server will not be restarted." 

But I can't even restart the server with "Restart server" command in VsCode, the same message will appear. There should be some way to manually restart the server again. Or what else am I suppossed to do?

**Comment by HighCommander4** [comment]
This suggests clangd is crashing on startup. When you perform "restart server", that does restart the server, it just crashes again on startup.

If you look at the Clangd section of the Output window, there should be a message indicating the reason for the crash. Feel free to post it here (or the entire log) for further diagnosis.

**Comment by Leon0402** [comment]
Hi, 

thank you very much, you are right, it actually just crashed constantly!

The reason it did was because it didn't detect a command line argument I parsed via the gui. 

I put into the field "--compile-commands-dir build" and clangd complained that it doesn't know that command line argument.It turned out, I have to put "--compile-commands-dir" and "build" into two seperate fields. 

Perhaps vscode-clangd could check if the passed params are valid and if not output the error message?

**Comment by Leon0402** [comment]
Just another question, I found another issue, but I'm not quite sure if it's a bug in clangd or in this extension.

![image](https://user-images.githubusercontent.com/33933142/93248673-7811b880-f790-11ea-9a34-c28fe4ad54e4.png)

It seems like doxygen documenation syntax is not recognized. Could you tell me, where I have to report it @HighCommander4 ?

**Comment by HighCommander4** [comment]
> Perhaps vscode-clangd could check if the passed params are valid and if not output the error message?

I don't think we want to duplicate the logic for validating clangd's command line parameters in both clangd and vscode-clangd (and all other clangd clients).

**Comment by HighCommander4** [comment]
> It seems like doxygen documenation syntax is not recognized.

Could you be more specific about what you are expecting / hoping to see?

**Comment by Leon0402** [comment]
> > Perhaps vscode-clangd could check if the passed params are valid and if not output the error message?
> 
> I don't think we want to duplicate the logic for validating clangd's command line parameters in both clangd and vscode-clangd (and all other clangd clients).

I was thinking more of just doing a test run perhaps? In the simplest way, you could just check if the server starts sucessfully.

**Comment by Leon0402** [comment]
> > It seems like doxygen documenation syntax is not recognized.
> 
> Could you be more specific about what you are expecting / hoping to see?

Here is the output of the C/C++ intellisense plugin: 
![image](https://user-images.githubusercontent.com/33933142/93249556-c3789680-f791-11ea-9551-c1056893b954.png)

As you can see it makes use of the \brief, \param options and displays the information in the right place (for example the param documentation next to the params)

**Comment by HighCommander4** [comment]
> I was thinking more of just doing a test run perhaps? In the simplest way, you could just check if the server starts sucessfully.

We do check that. And if it doesn't start successfully, we start it again. And if that happens 5 times we show the "crashed 5 times in the last 3 minutes" message.

(More precisely, I think it's vscode itself doing the above, rather than the vscode-clangd plugin.)

I think perhaps what you're suggesting is that, when this happens, instead of having to look in the Output window to see the reason for the crash, it would be better to show that in the dialog. I think that would be a useful improvement.

**Comment by HighCommander4** [comment]
> As you can see it makes use of the \brief, \param options and displays the information in the right place (for example the param documentation next to the params)

Thanks. I think this is a missing feature on the server side (it would have to parse the doxygen syntax and turn it into appropriate markdown for the client). Please feel free to file a [server-side issue](https://github.com/clangd/clangd/issues/new/choose) for it.

**Comment by HighCommander4** [comment]
> I think perhaps what you're suggesting is that, when this happens, instead of having to look in the Output window to see the reason for the crash, it would be better to show that in the dialog. I think that would be a useful improvement.

I'll re-title this issue to track this suggestion.

**Comment by Trass3r** [comment]
> But I can't even restart the server with "Restart server" command in VsCode, the same message will appear. There should be some way to manually restart the server again. Or what else am I suppossed to do?

I think that hasn't ever really worked for me, I always have to do Reload Window or it won't restart at all.
clangd just can't handle certain sources and crashes every time on them in my case, if you close the current editor this will help usually.

> > As you can see it makes use of the \brief, \param options and displays the information in the right place (for example the param documentation next to the params)
> 
> Thanks. I think this is a missing feature on the server side (it would have to parse the doxygen syntax and turn it into appropriate markdown for the client). Please feel free to file a [server-side issue](https://github.com/clangd/clangd/issues/new/choose) for it.

Probably there's a parser in clang already for `-Wdocumentation`.

> I think perhaps what you're suggesting is that, when this happens, instead of having to look in the Output window to see the reason for the crash, it would be better to show that in the dialog. I think that would be a useful improvement.

Or at least it should open the output window when clicking on the popup.

**Comment by tgsheng** [comment]
I have encountered this problem, and it appears frequently. What should I do? :sob:

**Comment by HighCommander4** [comment]
> I have encountered this problem, and it appears frequently. What should I do? sob

Look in the Output window (choose Clangd from the dropdown) for Clangd logs. If the cause of the crash / how to fix it is not clear from the logs, feel free to file an issue and attach the logs.
